### PR TITLE
[WIP] wasm-bindgen & ffi-gen

### PR DESCRIPTION
**Pull request by wngr**
changes to ffi-gen @ https://github.com/cloudpeers/ffi-gen/pull/60

to test
```
cd api
./build.sh
npx --yes http-server .
```
open browser and check console output:
```
RuntimeError: null function or function signature mismatch
    at <core::pin::Pin<P> as core::future::future::Future>::poll::hee1f14e768aa370f (05653dba:0xc3577a)
    at tlfs::api::FfiFuture<T>::poll::h76c2bf01723a3ed7 (05653dba:0x96beba)
    at tlfs::api::__create_memory_future_poll::{{closure}}::h662b2f4756108fd2 (05653dba:0x461870)
    at std::panicking::try::do_call::h2d6078dda0039457 (05653dba:0x87c3fc)
    at std::panicking::try::h92d29eb55d90adbf (05653dba:0x73cc07)
    at std::panic::catch_unwind::h3ccbc55a52d5dafb (05653dba:0xd1d2f9)
    at tlfs::api::panic_abort::hdc1653a7cab383da (05653dba:0x9225a4)
    at __create_memory_future_poll (05653dba:0xb40cef)
    at Api.createMemoryFuturePoll (bindings.mjs:655)
    at bindings.mjs:284
```

----

I worked around having to merge the `importObject` by providing the notifier callback via the global `window` object.

### CODE REVIEW

File: api/build.rs [hunk 1]
```diff
@@ -15,8 +15,8 @@ fn main() {
     let ffigen = FfiGen::new(&path).unwrap();
     let dart = dir.join("dart").join("lib").join("tlfs.dart");
     ffigen.generate_dart(dart, "tlfs", "tlfs").unwrap();
-    let js = dir.join("js").join("bindings.js");
+    let js = dir.join("js").join("bindings.mjs");
```
Comment (dvc94ch) [inline api/build.rs:18]: whats mjs?

File: Cargo.toml [hunk 2]
```diff
@@ -35,21 +35,36 @@ version = "0.41.0"
 default-features = false
 features = [
     "mdns",
+    "mplex",
```
Comment (dvc94ch) [inline Cargo.toml:38]: any particular reason why we need to support multiple muxers? I'm fine with either

Comment (wngr) [reply Cargo.toml]: nope, just copy-paste leftover

File: api/build.rs (hunk 1 above)
Comment (wngr) [reply api/build.rs:18]: So the thing should be imported as an ES module.

File: api/Cargo.toml [hunk 3]
```diff
@@ -12,10 +15,16 @@ default = ["capi", "futures"]
...
-ffi-gen = "0.1.5"
+ffi-gen = { version = "0.1.5", features = ["wasm-bindgen"] }
 
 [dependencies]
 anyhow = "1.0.51"
+ffi-gen = { version = "0.1.5" }
 ffi-gen-macro = "0.1.2"
 futures = { version = "0.3.17", optional = true }
 tlfs = { version = "0.1.0", path = ".." }
+tlfs-crdt = { path = "../crdt" }
```
Comment (dvc94ch) [inline api/Cargo.toml:26]: do we need this?

File: api/build-wasm.sh [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+#  wget -qO- \
+#  https://github.com/WebAssembly/binaryen/releases/download/version_$WASMOPT_VERSION/binaryen-version_$WASMOPT_VERSION-x86_64-linux.tar.gz \
+#  | tar xfz - binaryen-version_$WASMOPT_VERSION/bin/wasm-opt -O >> $WASMOPT
+#  chmod +x $WASMOPT
+#fi
+#
+#if [ "${OPTIMIZE}" ]; then
+#  echo "Optimizing wasm bindings with default optimization (this might take some time)"
+#  ./wasm-opt $OUT/local_first_bg.wasm -Oz -g --output $OUT/local_first_bg.opt.wasm
+#fi
```
Comment (dvc94ch) [inline api/build-wasm.sh:78]: what's with this stuff? do we need it?

File: api/js/index.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  try {
+    window.api = new Api();
+    const x = await init();
+
+    window.api.initWithInstance({ exports: x });
+    console.log(window.api.instance);
+
+    let pkg = Uint8Array.from(
+      atob(
+        "AAIDAAAAAAAAAAAAAAAAAAAAAAAABQAAAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAACAAAAOj///8AAAAAAAAAAAAAAAACAAAAdGl0bGUAAAUAAAAAAAAAAAgAAADo////AAAAAAAAAAAAAAAAAAIDAAAAAAAAAAAAAAAAAAAAAAAHAAAAdGl0bGUAAAXg////AAAAAAgAAADo////AAAAAAAAAAAAAAAAY29tcGxldGUCAAAACAAAAPT///8AAAAAAAAAAAgAAADo////AAAAAAAAAAAAAAAAY29tcGxldGUAAQAAAAAAAAAAAAAAAAAAAAAAAAcAAAAIAAAA4P///+D///8AAAAACAAAAOj///8AAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAIAAAB0aXRsZQAABQAAAAAAAAAABwAAAHRpdGxlAAAFpP7//wAAAAACAAAAdGFza3MAAAUAAAAAAAAAAAcAAAB0YXNrcwAABZD+//8AAAAABwAAAHRhc2tzAAAFpP7//wAAAAAHAAAAdGFza3MAAAW4/v//AAAAAAcAAAB0YXNrcwAABeD+//8AAAAABwAAAHRhc2tzAAAF/P7//wAAAAAHAAAAdGFza3MAAAUs////AAAAADj///8KAAAAdG9kb2FwcAcKAAAA/P3///gBAADs////AQAAAA=="
```
Comment (dvc94ch) [inline api/js/index.js:14]: hard coded binary blob?

File: src/lib.rs [hunk 6]
```diff
@@ -88,12 +100,28 @@ impl Sdk {
...
+                .unwrap()
+
+            //TODO
+            //        slf.add_external_address(
+            //            signaling_server
+            //                .with(Protocol::P2pWebRtcStar)
+            //                .with(Protocol::P2p(libp2p_peer.into())),
+            //            // TODO
+            //            AddressScore::Infinite,
+            //        )
```
Comment (dvc94ch) [inline src/lib.rs:115]: this needed or not?

File: src/lib.rs [hunk 7]
```diff
@@ -57,27 +55,41 @@ impl Sdk {
         .await
     }
 
-    #[cfg(not(target_family = "wasm"))]
+    #[allow(clippy::if_same_then_else)]
```
Comment (dvc94ch) [inline src/lib.rs:58]: clippy is annoying

**Review by dvc94ch — APPROVED** [review]

**Comment by dvc94ch** [comment]
would be nice to have a test for the native and web transports connecting, not sure how feasible that is

**Comment by dvc94ch** [comment]
why is the cloud relay commented out, is it not needed?

**Comment by wngr** [comment]
> would be nice to have a test for the native and web transports connecting, not sure how feasible that is

Got one of those for `libp2p-webrtc` here: https://github.com/wngr/libp2p-webrtc/blob/master/integration/src/main.rs

Will look into adding some tests here as well.

File: src/lib.rs (hunk 6 above)
Comment (wngr) [reply src/lib.rs:115]: Not sure yet. Will look into the actual libp2p connection behaviour next.

File: api/js/index.js (hunk 5 above)
Comment (wngr) [reply api/js/index.js:14]: Just something to get started. Will create a proper package soon.

File: api/build-wasm.sh (hunk 4 above)
Comment (wngr) [reply api/build-wasm.sh:78]: We should in order to reduce the final artifact's size.

**Pull request merged**